### ISSUE TITLE
undefined method `name' for nil:NilClass (NoMethodError): Image remove fails if vm contains cdrom

### ISSUE THREAD

**Issue by abbbi** [vagrant-libvirt/vagrant-libvirt#1262]
hi,

on a freshly setup centos8 stream with advanced virtualization and latest libvirt backend, the destroy operation on certain vagrant
domains fails with:

```
==> default: Removing domain...
/home/buildtest/.vagrant.d/gems/2.6.6/gems/vagrant-libvirt-0.4.1/lib/vagrant-libvirt/action/destroy_domain.rb:73:in `block in call': undefined method `name' for nil:NilClass (NoMethodError)
	from /home/buildtest/.vagrant.d/gems/2.6.6/gems/vagrant-libvirt-0.4.1/lib/vagrant-libvirt/action/destroy_domain.rb:72:in `select'
	from /home/buildtest/.vagrant.d/gems/2.6.6/gems/vagrant-libvirt-0.4.1/lib/vagrant-libvirt/action/destroy_domain.rb:72:in `call'
	from /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/lib/vagrant/action/warden.rb:48:in `call'
	from /home/buildtest/.vagrant.d/gems/2.6.6/gems/vagrant-libvirt-0.4.1/lib/vagrant-libvirt/action/prune_nfs_exports.rb:32:in `call'
	from /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/lib/vagrant/action/warden.rb:48:in `call'
	from /home/buildtest/.vagrant.d/gems/2.6.6/gems/vagrant-libvirt-0.4.1/lib/vagrant-libvirt/action/forward_ports.rb:199:in `call'
	from /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/lib/vagrant/action/warden.rb:48:in `call'
	from /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/lib/vagrant/action/builtin/provisioner_cleanup.rb:25:in `call'
```

the Virtual machine itself seems to be removed, but the volume is leftover:

```
virsh vol-list --pool cache | grep rhel8-rts
 rhel8-rts_default.img           /cache/rhel8-rts_default.img
```

Seems there is some issue with detecting the volume name that is to be removed.

The cause for the issue is a defined CDROM device:

- CDROM(hda):        /local_iso/rhel-8.3-x86_64-dvd.iso

Removing the CDROM device from the vagrant VM definition makes the detroy work.

**Comment by electrofelix** [comment]
Looks like a duplicate of #1209

**Comment by electrofelix** [comment]
I thought the original issue would be fixed with the merge of https://github.com/fog/fog-libvirt/pull/75 and release of fog-libvirt-0.8.0.

Can you check what version of fog-libvirt is running? I'll have to see if I can do a test locally to see what it is returning, maybe we need to just check for nil results and skip instead of trying to avoid receiving nil results.

**Comment by abbbi** [comment]
It seems hat during destroy, the fog instance returns an empty volume, i could make
it work with this simple workaround:

x.name == libvirt_domain.name + '.img' if x

in lib/vagrant-libvirt/action/destroy_domain.rb:73

then everything works flawlessly.

.vagrant.d/gems/2.6.6/gems/fog-libvirt-0.8.0/

**Comment by electrofelix** [comment]
@abbbi that doesn't look like it's a nil volume getting returned? If you run irb from vagrant embedded it should be possible to do a more detailed debug to see what is happening compared to what I'm seeing locally which suggests this bug should be fixed (as in I can't reproduce with a VM with a cdrom attached).

My Vagrantfile
```
Vagrant.configure("2") do |config|
  config.vm.box = "generic/debian10"

  config.vm.provider 'libvirt' do |lv|
    lv.storage :file, :device => :cdrom, :path => "#{ENV['HOME']}/Downloads/rhel7.iso"
  end
end
```

After bringing up, running irb from the same directory as the Vagrantfile:
```
require 'fog/libvirt'
conn_attr = {:provider => 'libvirt', :libvirt_uri => 'qemu:///system'}
conn = Fog::Compute.new(conn_attr)
domain = conn.servers.get(`cat .vagrant/machines/default/libvirt/id`)
domain.volumes
```

This returns the following for me:
```
 => [  <Fog::Libvirt::Compute::Volume
    id="/var/lib/libvirt/images/vagrant-libvirt_default.img",
    pool_name="default",
    key="/var/lib/libvirt/images/vagrant-libvirt_default.img",
    name="vagrant-libvirt_default.img",
    path="/var/lib/libvirt/images/vagrant-libvirt_default.img",
    capacity=128,
    allocation=0,
    owner="0",
    group="0",
    format_type="qcow2",
    backing_volume=nil
  >,   <Fog::Libvirt::Compute::Volume
    id=nil,
    pool_name="default",
    key=nil,
    name="fog-805336606807416",
    path=nil,
    capacity="10G",
    allocation="1G",
    owner="0",
    group="0",
    format_type="raw",
    backing_volume=nil
  >]
```

If I run the following as well around the area the code failed for you:
```
libvirt_domain = conn.client.lookup_domain_by_uuid(`cat .vagrant/machines/default/libvirt/id`)
domain.volumes.select do |x|
    x.name == libvirt_domain.name + '.img'
end.first
```

It outputs the following:
```
<Fog::Libvirt::Compute::Volume
    id="/var/lib/libvirt/images/vagrant-libvirt_default.img",
    pool_name="default",
    key="/var/lib/libvirt/images/vagrant-libvirt_default.img",
    name="vagrant-libvirt_default.img",
    path="/var/lib/libvirt/images/vagrant-libvirt_default.img",
    capacity=128,
    allocation=0,
    owner="0",
    group="0",
    format_type="qcow2",
    backing_volume=nil
>
```

I'd just like to confirm there isn't something else going on as well here. Because the above aligns with my understanding of what was previously happening, while the clear failure you've encountered indicates that understanding is incomplete, and there might be other areas that need to be checked as well.

**Comment by electrofelix** [comment]
oh, wait, it looks like I have fog-libvirt 0.7.0 installed, I wonder if something has happened to re-introduce the bug

**Comment by electrofelix** [comment]
And we have a winner, it seems I managed to re-introduce the bug when attempting to fix something else...

**Comment by electrofelix** [comment]
So it looks like my code is only partially responsible for the bug, previously 0.7.0 would return both disks and cdroms, 0.8.0 now returns a nil volume instead of the cdrom. The bug I fixed ensured that for single entries a nil entry would be returned instead of a hash, so essentially anywhere we use select on volumes, we should filter to ensure they are not nil before attempting to access, similar to your workaround.

If you can get to it before me, the fix should resolve both this issue and #1209
